### ISSUE TITLE
BUG: training ddp fraction

### ISSUE THREAD

**Issue by johnnynunez** [ultralytics/ultralytics#9785]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

In DDP mode, i don't now if it is with the random split.
it causes in DDP Mode with fraction: 0.5
```bash
assert cache["hash"] == get_hash(self.label_files + self.im_files)  # identical hash
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@johnnynunez hello! Thanks for reporting this issue. 🙏 It seems like there might be a mismatch or an update issue with your dataset cache when using Distributed Data Parallel (DDP) mode with a fraction less than 1.

One preliminary solution could be to try clearing your dataset cache. This can be done by removing the cache files (usually ending with '.cache') in your data directory and then rerunning your training command.

```bash
# Example to remove cache files in your data directory
rm -rf /path/to/your/data/*.cache

# Rerun your training command
```
Remember to replace `/path/to/your/data/` with the actual path to your dataset.

If this doesn't solve the issue, it might require a deeper look into how the dataset fraction is handled in DDP mode. Could you please share additional details like how your dataset is structured or if any modifications were made to the training pipeline?

I appreciate your offer to contribute! While we investigate this further, any additional details or observations you can provide would be invaluable. 🚀

**Comment by johnnynunez** [comment]
> @johnnynunez hello! Thanks for reporting this issue. 🙏 It seems like there might be a mismatch or an update issue with your dataset cache when using Distributed Data Parallel (DDP) mode with a fraction less than 1.
> 
> One preliminary solution could be to try clearing your dataset cache. This can be done by removing the cache files (usually ending with '.cache') in your data directory and then rerunning your training command.
> 
> ```bash
> # Example to remove cache files in your data directory
> rm -rf /path/to/your/data/*.cache
> 
> # Rerun your training command
> ```
> Remember to replace `/path/to/your/data/` with the actual path to your dataset.
> 
> If this doesn't solve the issue, it might require a deeper look into how the dataset fraction is handled in DDP mode. Could you please share additional details like how your dataset is structured or if any modifications were made to the training pipeline?
> 
> I appreciate your offer to contribute! While we investigate this further, any additional details or observations you can provide would be invaluable. 🚀

I think that the seed maybe it's no the same for all gpus when splits the data. Maybe it is necessary to do a broadcast

**Comment by glenn-jocher** [comment]
Hey @johnnynunez, thanks a lot for the follow-up and the insight! 😊 It sounds like synchronizing the seed across all GPUs could indeed help with the data split consistency in DDP mode. You could try setting a manual seed right before your dataset gets split. Here's a quick way to ensure the same seed is used across all your GPUs:

```python
import torch

# Ensure the same seed across all processes
torch.manual_seed(42)
```

Remember to place this before the dataset split operation. This step helps maintain dataset consistency across different processes.

Let us know how this goes, and if the issue persists, we're here to dive deeper. Thanks for helping to improve YOLOv8!

**Comment by glenn-jocher** [comment]
@johnnynunez after reviewing this I've reverted the `fraction` behavior back to it's original state in https://github.com/ultralytics/ultralytics/pull/9851/commits/2a22d0bd662ab1e0ef99f3adba4c632979d9ebc3. It will now load the first fraction of the dataset without randomly sorting the indices of the images it loads.

This should resolve any possible DDP issues.
